Improve TowerResolver performance, especially for REPL

ileasile opened the pull request:
Implemented definitelyDoesNotContainName in some scope implementations and added special annotation for skipping extensions resolving for implicits. Compiling times are now increasing lineary that allows to compile hundreds and thousands of snippets without serious performance degradation.

ligee commented on compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt:
```diff
@@ -227,9 +229,16 @@ class TowerResolver {
-                TowerData.BothTowerLevelAndImplicitReceiver(nonLocalLevel, implicitReceiver).process()?.let { return it }
-            }
+            val receiverValue = implicitReceiver.receiverValue
+            val resolveExtensions = receiverValue !is ImplicitClassReceiver ||
```
The intentions will look much cleaner if you'll reverse the flag meaning, i.e. name it `skipResolveExtensions` and rewrite the code accordingly.

ligee commented on libraries/scripting/common/src/kotlin/script/experimental/api/scriptCompilation.kt:
```diff
@@ -148,7 +148,7 @@ val ScriptCompilationConfigurationKeys.refineConfigurationOnAnnotations by Prope
 
 /**
- * The list of script fragments that should be compiled intead of the whole text
+ * The list of script fragments that should be compiled instead of the whole text
```
Please, to the separate commit (with the "[minor]" prefix), it is completely irrelevant here.

ileasile replied:
Done

ligee commented on plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt:
```diff
@@ -265,11 +265,18 @@ open class ReplCodeAnalyzerBase(
+            val combinedLexicalScopes = if (hasImports)
+                lexicalScopeAfterLastLine.replaceImportingScopes(scopesForThisLine.importingScope)
+            else
+                lexicalScopeAfterLastLine
```
Have you checked that this does not affect implicit imports, that are passed via script compilation configuration?
I believe it shouldn't, but it must be checked, I believe.
You'll need to construct a REPL with a custom script def with implicit imports, to check that. Maybe it makes sense to add tests for it too, we have not many tests with imports and this change can alter the behavior.

ileasile replied:
@ligee are you talking about `defaultImports`, right?

ileasile replied:
If so, I've a test

dzharkov commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/AbstractLazyMemberScope.kt:
```diff
@@ -188,6 +188,12 @@ protected constructor(
+    override fun definitelyDoesNotContainName(name: Name): Boolean =
+        !getVariableNames().contains(name) &&
+                !getFunctionNames().contains(name) &&
+                getClassifierNames()?.contains(name) != true &&
```
Sorry, it's not obvious but this is incorrect semantic usage of the `getClassifierNames()`: it returns null when the scope can't compute classifier names' set efficiently.
And this happens to `LazyClassMemberScope` unfortunately.
It's necessary to implement it properly there to make `definitelyDoesNotContainName` work correctly

ileasile replied:
@dzharkov Fixed this method, please review new implementation

ileasile commented:
@ligee @dzharkov I've deleted annotation and related fixes in `TowerResolver`, but added one small thing,
```
override fun definitelyDoesNotContainName(name: Name): Boolean = true
```
for **empty** member scope. It seems to make linear complexity in the test I've added. Maybe it will even suit us in Kotlin kernel, so I ask you to review and merge it.

dzharkov commented on core/descriptors/src/org/jetbrains/kotlin/resolve/scopes/MemberScope.kt:
```diff
@@ -46,6 +46,8 @@ interface MemberScope : ResolutionScope {
             p.println("Empty member scope")
         }
 
+        override fun definitelyDoesNotContainName(name: Name): Boolean = true
```
What is this change for?

ileasile replied:
There are empty member scopes created in REPL scripts, so it helps us to filter them out

dzharkov commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt:
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
+                return false
+            if (getSyntheticFunctionNames(thisDescriptor).contains(name))
+                return false
+            if (getSyntheticNestedClassNames(thisDescriptor).contains(name))
```
Also, I'm concerned we don't handle usual nested classifiers' names here

ileasile commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt:
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
     override fun getFunctionNames() = _functionNames
 
+    override fun definitelyDoesNotContainName(name: Name): Boolean {
+        if (getVariableNames().contains(name) || getFunctionNames().contains(name))
```
Fixed

ileasile commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/lazy/descriptors/LazyClassMemberScope.kt:
```diff
@@ -135,6 +135,27 @@ open class LazyClassMemberScope(
+        with(c.syntheticResolveExtension) {
+            if (getSyntheticCompanionObjectNameIfNeeded(thisDescriptor) == name)
+                return false
+            if (getSyntheticFunctionNames(thisDescriptor).contains(name))
```
Fixed
